src/ips.rs

```diff
@@ -0,0 +1,445 @@
+                    IpAddr::from_str(nodes[i].ip.as_str()).unwrap(),
+                    IpAddr::from_str(nodes[j].ip.as_str()).unwrap(),
+                );
+                graph.insert(edge);
+            }
+        }
+        graph
+    }
+
+    fn determine_degrees_factors(&mut self, degrees: &HashMap<IpAddr, u32>) {
+        let min = *degrees
+            .iter()
+            .min_by(|a, b| a.1.cmp(b.1))
+            .map(|m| m.1)
+            .unwrap_or(&0);
+        let max = *degrees
+            .iter()
+            .max_by(|a, b| a.1.cmp(b.1))
+            .map(|m| m.1)
+            .unwrap_or(&(NORMALIZE_TO_VALUE as u32));
+
+        self.degree_factors = NormalizationFactors {
+            min: min as f64,
+            max: max as f64,
+            value: NORMALIZE_TO_VALUE,
+        };
+    }
+
+    fn determine_eigenvalue_factors(&mut self, eigenvalues: &HashMap<IpAddr, f64>) {
+        let min = *eigenvalues
+            .iter()
+            .min_by(|a, b| a.1.partial_cmp(b.1).unwrap())
+            .map(|m| m.1)
+            .unwrap_or(&0.0);
+        let max = *eigenvalues
+            .iter()
+            .max_by(|a, b| a.1.partial_cmp(b.1).unwrap())
+            .map(|m| m.1)
+            .unwrap_or(&NORMALIZE_TO_VALUE);
+
+        self.eigenvector_factors = NormalizationFactors {
+            min,
+            max,
+            value: NORMALIZE_TO_VALUE,
+        };
+    }
+
+    fn determine_betweenness_factors(&mut self, nodes: &[Node]) {
+        let min = nodes
+            .iter()
+            .min_by(|a, b| a.betweenness.partial_cmp(&b.betweenness).unwrap())
+            .map(|m| m.betweenness)
+            .unwrap_or(0.0);
+        let max = nodes
+            .iter()
+            .max_by(|a, b| a.betweenness.partial_cmp(&b.betweenness).unwrap())
+            .map(|m| m.betweenness)
+            .unwrap_or(NORMALIZE_TO_VALUE);
+
+        self.betweenness_factors = NormalizationFactors {
+            min,
+            max,
+            value: NORMALIZE_TO_VALUE,
+        };
+    }
+
+    fn determine_closeness_factors(&mut self, nodes: &[Node]) {
+        let min = nodes
+            .iter()
+            .min_by(|a, b| a.closeness.partial_cmp(&b.closeness).unwrap())
+            .map(|m| m.closeness)
+            .unwrap_or(0.0);
+        let max = nodes
+            .iter()
+            .max_by(|a, b| a.closeness.partial_cmp(&b.closeness).unwrap())
+            .map(|m| m.closeness)
+            .unwrap_or(NORMALIZE_TO_VALUE);
+
+        self.closeness_factors = NormalizationFactors {
+            min,
+            max,
+            value: NORMALIZE_TO_VALUE,
+        };
+    }
+
+    fn rate_node(&self, node: &Node, degree: u32, eigenvalue: f64) -> f64 {
+        // Calculate rating for node
+
+        // Rating is a combination of the following factors:
+        // 1. Degree
+        let mut rating = ((degree as f64 - self.degree_factors.min)
+            / (self.degree_factors.max - self.degree_factors.min)
```

Comment:
`struct NormalizationFactors` could have a `lerp` method, that handles the linear interpolation.

Reply:
Added.